### ISSUE TITLE
[bug] Getting req.setLocale is not a function from server

### ISSUE THREAD

**Issue by IagoSRL** [loconomics/loconomics-backend#7]
## Reproduce
- On Windows 10, branch `master`.
- Run server with `npm run dev`
- Try to access an API, like in the browser console: `fetch('http://localhost:1337/api/v1/en-US/states-provinces').then((a) => console.log(a)).catch((b) => console.error(b));
- Got next errorr
```
(node:39080) UnhandledPromiseRejectionWarning: TypeError: req.setLocale is not a function
    at Object.<anonymous> (D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\i18n.ts:9:11)
    at step (D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\i18n.ts:32:23)
    at Object.next (D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\i18n.ts:13:53)
    at D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\i18n.ts:7:71
    at new Promise (<anonymous>)
    at __awaiter (D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\i18n.ts:3:12)
    at module.exports (D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\i18n.ts:3:31)
    at Layer.handle [as handle_request] (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\sails\node_modules\express\lib\router\layer.js:95:5)
    at trim_prefix (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\sails\node_modules\express\lib\router\index.js:317:13)
    at D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\sails\node_modules\express\lib\router\index.js:284:7
    at Function.process_params (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\sails\node_modules\express\lib\router\index.js:335:12)
    at next (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\sails\node_modules\express\lib\router\index.js:275:10)
    at D:\Iago\Proxectos\Loconomics\loconomics-backend\api\middleware\auth.ts:33:5
    at allFailed (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\passport\lib\middleware\authenticate.js:107:18)
    at attempt (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\passport\lib\middleware\authenticate.js:180:28)
    at Strategy.strategy.fail (D:\Iago\Proxectos\Loconomics\loconomics-backend\node_modules\passport\lib\middleware\authenticate.js:297:9)
(node:39080) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting
a promise which was not handled with .catch(). (rejection id: 7)
```

It blames that `req.setLocale` is not a function, while is documented at Sailsjs website https://sailsjs.com/documentation/reference/request-req/req-set-locale

**Comment by ndarilek** [comment]
Hey, sorry, I'd recommend trying out the loconomics-data branch. I've 
mostly completed a migration to @loconomics/data there, and I'm pretty 
sure this shouldn't be an issue there. :)


Thanks, hoping to have that finished and merged over the next couple weeks.

**Comment by IagoSRL** [comment]
Saw this message now, so forget last question at Slack; will check that branch next week, thanks.

**Comment by IagoSRL** [comment]
This error seems to not throw at the loconomics-data branch, but cannot confirm completely because I'm having problems to start the server with the proper database connectionstring to my local (it's a named instance, gives problems with tedious).

When I run the app, it takes 3-4 minutes to start-up Sails and I get a 15 seconds timeout trying to connect database. The timeout is normal if not able to connect with the given settings, but the extra time is not and not sure why, but is making really slow to test different set-up combinations. I keep trying

**Comment by IagoSRL** [comment]
I still didn't fix this, but went into debugging the server and found what seems the cause of slow startup: there are a lot of exceptions being throw (but catched) regarding loading files that doesn't exist, mostly with extensions `.tsx, .ts, .js`. Most from inside the `tpeorm` package, but I didn't check everything since are thousands of files. It was trying to load Typescript definitions too (.d.ts).

Seems all related to typeorm and ts-node. This makes start-up extremely slow on my end. And it burns the CPU while on this initialization task (if that happens at production, sure any autoscale set-up will scale it to more instances, even if that ones are not needed since the high load is only on initialization).
I understand that `ts-node` module resolver will try to locate `ts` files first (not sure why try `.tsx` though), but I would expect that almost one file for each extension group would exists, but seems to try loading files that don't exist on any extension. One example I remember, is `node_modules/typeorm/reflect-metadata.ts`.

Not sure if some version problem or just the expected behavior, but it makes booting really slow.

**Comment by ndarilek** [comment]
Hmm, interesting, I don't see this at all. Granted, the app runs in 
Docker containers for me, but that shouldn't matter regarding that. Some 
questions:


  * Does this happen when you remove the node_modules directory and 
start from scratch?

  * What steps are you using to bootstrap the backend from an empty 
directory?

  * Are you using Yarn or NPM?

  * How are you running the app?


Maybe this is something unique to Windows/NTFS. I definitely wouldn't 
ship something that takes minutes to start. :)

**Comment by ndarilek** [comment]
So I'm running the following Bash script on my system. As can be seen, 
this sequence of commands doesn't run the new backend in Docker, so you 
should be able to run them on your Windows system with little modification:


```bash

#!/bin/bash

git clone --recursive --branch loconomics-data 
https://github.com/loconomics/loconomics-backend tmp
pushd tmp
yarn
MSSQLSERVER_URL=mssql://sa:Password!!11@localhost:1433/loconomics yarn dev
popd

```


The script fails to authenticate to the database because I've created a 
fresh instance for this test without the loconomics data imported, but 
it gets to the authentication/failure step in around 30 seconds--and 
that's to clone the repo, do the `yarn install`, etc. And the production 
environment is Docker-based, so it should perform equally well there.


Anyhow, this sounds like something different from the initial 
`req.setLocale` issue, so I'm going to close this and invite you to open 
a new issue if the problem persists. But I don't know how much help I 
can be given that I can't replicate it and don't have a lot of time 
remaining. I know Docker may be a new and unfamiliar tool in your 
workflow, but it generally irons out platform differences like these, 
and part of why I use it so heavily is so we can have a single blessed 
platform that everyone works with.

**Comment by IagoSRL** [comment]
Agree with this going off-topic; didn't close it myself since I wanted to confirm the reported on the latest branch, which I still couldn't try, but I suppose is not happening.

About times, I know by experience that NTFS and Windows filesystem API in general is a lot slower than the Linux one, then is not surprising it boots faster at Linux. Still, for a server that does a few things rigth now, it's still a lot what I'm getting, even if was just 10 seconds, since I have a complex complete server on Nodejs that takes around 2 seconds to boot up on Windows on this same machine, same underlying SSD unit (I measured it now since I never had the need for it before -is fast enough). I think the major problem is `ts-node` file lookups and on-fly compilation, plus some from Sailsjs. Not sure how this will scale.

@ndarilek I know you have not time to assist me on this, no problem, I understand completely. I publish all this details just for the record:

### Set-up and running
I did a fresh install last week, but I repeated just in case with latest content from branch `loconomics-data` and removing node_modules, using Yarn after.
From Bash I run `node app.js`.

### First trial
I got next error after 2:38 elapsed time (truncated output).

```
error: A hook (`helpers`) failed to load!
error: Failed to lift app: Error: Attempted to `require('D:\Iago\Proxectos\Loconomics\loconomics-backend\api\helpers\mssql.ts')`, but an error occurred:
....
error: A hook (`orm`) failed to load!
```

It has not settings for database connection, so it fails early. It's interesting because it shows some kind of initial set-up of the server or Sailsjs, double checking the configuration is in place before attempt anything else, so it lets me know when ts_node ended loading modules and starts executing them. I did a second attempt just to compare times, with less than a second of difference.

### Second trial
I defined the `MSSQLSERVER_` environment variable with my local settings and tried again. Like the other day, is not able to connect to my local SQL Server instance (remember that the mix of the underlying driver, tedious, plus the use of environment variables and slash character scaping on Bash for Windows, causes problems with "named instances" like the one I have for development with all the standard defaults from MS SQL Server Express --the name of the dev version from some years ago). The syntax for host name and named instance is `localhost\SQLEXPRESS`, while on env variables I have to escape as `localhost\\SQLEXPRESS`, still giving problems (but almost is not trying to connect to wrong host `localhostSQLEXPRESS` as happens when specifying just one back slash). I tried with the machine name rather than localhost too (ESTUDIO-I3 on my case).

Going to error an times, it first shows `UnhandledPromiseRejectionWarning: ConnectionError: Port for SQLEXPRESS:1433 not found in estudio-i3` after 2:42 minutes. After the error stack and some seconds after shows the 'info' messages from Sailsjs boot-up, including the ship draw with ASCII art, confirming the server was lifted, environment and port in use. After that, 15 seconds later a new error appear similar to first one for database error connection `UnhandledPromiseRejectionWarning: ConnectionError: Failed to connect to ESTUDIO-I3:1433 in 15000ms`.
The server is up and nothing more gets in the output if requests don't happen. I understand this is the moment the boot process has ended so is around 3 minutes overall, but obviously will be less without the database problems because of the 15 seconds timeout. It's not clear from the first error message if that first attempt has the same timeout, I will guess yes which means the complete success process should take around 1:27.

### Third attempt
I set my database set-up at the defaults of the config file, as a  hack just to prevent problems mixing up with environment variables. But doesn't get the config for some reason (even if there is an example of the file, the example doesn't works). So same result as first attempt.

### Four attempt
Ubuntu Linux on Windows through WSL (Windows Subsytem for Linux).
It's much faster booting up and giving the connectivity error, just 41 seconds! I expected something better because how Linux kernel handles file stats (done a lot by ts-node module loader checking if each module to be loaded exists as a `.ts` file `.js` file or even other extensions), but because the underlying disk and file system is the same I expected less differences, so this result is quite impressive. Maybe the implementation details of Nodejs 'fs' module for Windows adds something. Under a real Linux system with ext3/4 filesystem I expect even better results.

I used the environment variable on this, and the first connectiviy error has gone. The first output is the Sailsjs output (the ship draw, server lifted and info). But I get a database connection error after that, and is like the second error I got in other attempts but with some extra info about Node deprecating unhandled exceptions (the reason for this should be different Nodejs version on this environment). This has the 15 seconds timeout, so the real time should be 15 seconds faster on this environment. On any case, the code that throws this starts at `Timer.processTimers`. It's strange it gives only one connection error here, like if getting the settings for one of this fails and works for the other one, but that's just an initial guess. Trying to connect to the server fails cause it keeps waiting without response. I tried different combinations for the connection string without luck.

### Final thoughs
I guess the connectivity problem comes with using a named instance of SQL Server (that is the default setting) and how the driver manages that in the connection string. Cannot try it now because of further problems with the set-up forbide me from installing another instance.

I will recommend for anyone working on Windows on this to use a default name instance, and use the Ubuntu Bash (WSL) since is much faster. The problem may come when needing to use the debugger.

Too, in order to use Docker, a "Pro" edition of Windows is recommended, since it allows to use the Docker set-up with the proprietary Windows-native Hyper-V virtualization and networking, that should be better than the alternative that uses VirtualBox under the hood, and has some performance and set-up problems (for example, it behaves like a different machine so will not react to 'localhost'). But I wasn't able to try it (only the VirtualBox thing, and did not worked for the project set-up).

**Comment by ndarilek** [comment]
Got it. To be clear, I'm happy to help troubleshoot, it'd just be 
difficult for me to run everything in Windows myself. So a few notes:


1. Speed of compiling .TS files: on my system this is very fast, so I'm 
not sure what's going on under Windows. The production system I set up 
runs everything in a Docker image based on Alpine Linux, so I would 
assume similar performance in production. In any case, the actual 
package name of these files is @loconomics/data. I use a bit of hackery 
to convert those references to the data/ directory, so if it ever proves 
slow, it should be easy to publish pre-built packages and undo the 
reference, or come up with a way to use it only in development mode.


2. mssql.ts: That was an old helper used to obtain a database connection 
from a pool. You can still see it in use in the endpoint 
api/controllers/users/profile.ts. I'm hoping to port that endpoint by 
EOW, so ideally that helper will go away then. It's odd that this fails 
for you, but only when you're not using WSL. I've only seen the first 
part of the load failure, so it's hard for me to figure out what might 
be going wrong. Could you please send the *entire* traceback for these 
exceptions so we can figure out what's going wrong?


3. WSL: Glad this works better. I wouldn't expect performance on par 
with a full Linux system since WSL's EXT4 emulation is slower. What I'm 
more interested in is why some modules fail to load completely under 
WSL. The filesystem/emulation layer doesn't seem like it should just 
block a module from compiling.


4. Docker: had no idea it was this complicated under Windows. I'm sorry. 
It's hard to test everything under multiple platforms given the time 
constraints I've been under, and Docker at least ensures that my local 
environment is identical to staging/production. I prioritized that over 
spending time testing under Windows.


5. Named database instances: I'm not sure this is compatible with how 
I'm doing configuration at the moment. Under environments like Heroku, 
K8s, Docker Swarm, etc. it is traditional to inject database 
configuration as an environment variable, and URLs are fairly standard 
for configuring Postgres/MySQL/MariaDB. I don't think the issue is your 
escaping, but rather the fact that "localhost\\whatever" isn't a valid 
hostname when the URL is parsed from the environment. It isn't an 
insurmountable problem, but I optimized for the case of getting 
something working in Docker and making it easy to inject a different 
database configuration when the image goes to production. I'd never 
heard of this concept of distinguishing database instances based on 
appending "\name" to the hostname except for briefly in passing a couple 
months back, so didn't know its implications on running it in your 
development setup. To see what I experimented with, you can run the 
following in Node:


```

 > url.parse("http://\"localhost\\foo\"")
```


I suspected that wouldn't work and it didn't. If there's some way to 
trick the URL parser into accepting a hostname with backslashes then 
that might work, but I imagine the URL-parsing logic is pretty strict 
and standard.


I'm sorry. I wish we'd worked together on this before my last week on 
this project. As it is, I'm trying to exhaustively document all the 
patterns I've been using so folks can move forward with what they've 
done, so I probably don't have any additional time to debug this. If you 
want to support named instance development, you'll probably want to 
investigate TypeORM's configuration. Maybe you can provide the 
individual configuration options as separate environment variables, or 
take them from another source. I just felt that the URL was very 
concise, and made it easy to completely change the database 
credentials/location based on the environment. It looks like the longer 
form of TypeORM's connection options should do what you'll need--you'll 
just need to specify the host, port, username, password, etc. as 
individual configuration items for each environment.

**Comment by IagoSRL** [comment]
@ndarilek I know you are open to help, sorry if my comment was easy to misunderstood! My point is that I understand our timing limitations, don't worry.

Some comments on your points for clarification:
- 1. I "suppose" the major problem is the ts-node module loader, that for every `require` line that has no extension (usually all, as is the standard in Nodejs projects) tries to locate a matching file looking for the different extensions supported. For example for a `require('./main')` it will run a file stat for `./main.ts`, `./main.tsx`, `./main.d.ts`, `./main.d.js` and load the first file that exists. Maybe there are more extensions in the lookup, and that adds to the directory index file lookup that Nodejs does by default. This comes from a design error of Nodejs of making extensions optional (in recent words of its creator) and with `ts-node` get incremented a lot because the additional extensions. If the system is not too fast on checking file existence, it degrades quickly, as I "think" is the case at Windows.

- 2. Understood, I will wait that fix. No need for additional trace but basically was a connection error because cannot understand the connection string.

- 3. For clarification, the errors I got here were the same as without WSL, connection to database. It's just that with same settings it showed one rather than two of the errors, like if gets the setting correctly for one of the database connections but not for the other (which seems to not make sense). Anyway, may get fixed by 2.

- 4. I agree. Just for the record I will add that Docker is technically more complex on any non-Linux platform, because it's design is tighed to a couple of Linux specific, "propietary", features. It's not an UNIX or POSIX technology, not part of a standard and therefore not available at others kernels, even in the Unix world. Docker wasn't expected to be cross platform, but most like "cross distro". However, to make it available to others, they add additional sofware in the package called "Docker for Desktop" that creates a full virtual machine with a custom minimal Linux distro; that VM is the one that runs Docker; additional artifacts are added so Docker can appear to be part of the host OS and automatically connected to the docker CLI commands at the host system. They look to me like a set of workarounds to be honest, even if officially supported. The one that uses VirtualBox at Windows is far from perfect, since the performance is poor and is not able to look as localhost, is forced to be like another device on the network (and so, any image must open firewall rules to all the network or will be as isolated as is not reachable), plus other problems; it's not recommended right now. Alternatively, on Windows Server and Pro editions, it uses HyperV rather than VirtualBox, fixing some of that issues. On any case, the CPU must support hardware virtualization, except on real Linux of course where no VM is involved. Anyway, Docker runs ever on top of a VM when run out of a Linux host system, so all usual complications apply.

- 5. URLs should work since I found at the `tedious` source code a custom URL parsing just to support the concept of "named instances". At some point working at the `loconomics-data` I was able to make it work, after lot of trials, but same tricks seems to not work here for some reason or I forgot some detail. I think that the whole concept of named instances it just to abstract the need to define custom ports when several SQL Server instances runs on the same machine (cause of different versions, or created by desktop software transparently for the user). A service/daemon listen for the standard port and redirects connections to every running server based on name (or without name, is redirected to the 'default instance'). I think is a wrong design that adds complexity since deviates from any standard, but don't know all the requirements that lead to that; it seems we have to deal with it; at the Linux version seems MS ditched the instance name, almost as default value, which is good, but on Windows it applies a default name and even a name is required for development or non-server editions (not sure if at all versions or if that changed on latest ones). A part of that, I tried individual environment variables too rather than URL without luck.

--------------------

Thanks @ndarilek for all the documentation.

**Comment by ndarilek** [comment]
Just so I understand, you aren't currently running Windows 10 Pro?


I get that running Docker requires virtualization of some sort, but in 
my experience, any modern X86-64 CPU has this support. For reference, 
the CPU in my 10-year-old Dell Latitude E4300 has virtualization 
capabilities. If the expectation is that someone will run SQLServer 
locally for development, their CPU will almost by necessity be modern 
enough to support this feature. And yeah, the resulting containers may 
not be running on localhost if the Docker instance is using the older 
toolbox, but they should be available at an IP like 10.0.2.2 or 
something similar. At one point I successfully pointed Visual Studio 
2017 at a containerized SQLServer instance running in Docker, so using a 
different IP shouldn't introduce significant complications at this scale.


I understand that Docker is a new tool for both of you, and respect that 
changing one's workflow is a big request. At the end of the week this 
project will be entirely yours, and the choice to continue supporting 
Docker or not will be out of my hands. As someone who *himself* 
significantly changed his workflow to develop for Loconomics, I'd 
*strongly* advise trying to learn Docker, and would recommend against 
not changing the developer experience significantly because it doesn't 
work well with proprietary or custom configurations outside of Docker. 
There is already a lot of complexity in this project, and while Docker 
adds more, it hides away much of what complexity already exists in order 
to make the development environment nearly identical to production. 
Essentially, the codebase in Docker should work identically on my Linux 
desktop, on @joshdanielson's Mac, and in production. It already worked 
well on Trevor's system as well, which I think may also have been a mac. 
The only place it *isn't* working is on your system outside of Docker, 
and while nothing in the current setup is specific to Docker, the goal 
with that choice was to minimize the impact that different system 
configurations would have on a development process that is already 
pretty time-constrained.


TLDR: I've tried not to be pushy about adopting/changing workflows I 
think, but I'd like to be a little pushy about this one. :) If you want 
a system that's easy for potential volunteers to help out with, without 
them first having to install SQLServer/a specific version of 
Node/TypeScript/bunches of other things, I'd really recommend giving 
Docker a shot. I understand it's a new thing to learn, but as someone 
who spent much of his time with you guys learning new things, I think 
it's well worth the tradeoff. :) And with that I'll let this go and 
leave the decision to you.

**Comment by IagoSRL** [comment]
I'm not running Win 10 Pro. I upgraded to 10 from Win 7 Business, but Microsoft had that funny thing to upgrade only to the 'basic' version, not the equivalent one. Anyway, I had no need for pro until now, but I'm looking if I have to switch to it.

I think Docker is great, just pointing that Docker promise of no differences running an instance at different platforms is far from exact, despite their marketing speech. I'm learning it and I understand is becoming a mandatory skill in our industry, and while able to running it on Linux following docs and tutorials, at Windows the story is being different for me (and this is my main system right now). I had the toolbox problems last week with different connection and IP and I was able to run the Docker instance for this project but unable at all to connect to it through `http`, every attempt was refused. I tried several advices to solve it from the internet, without luck.
I have much more to learn about Docker, for sure! And I want. But time pressure is a big issue for me.

I'm trying to change my workflow (as I keep doing usually; I did it when entered the project, too, later when we changed a large part of the architecture and target, and even based on several contributors with new proposals here and there). It's part of our work, and there is friction and lack of experience on every iteration.

About SQL in Docker, I think that's a great tool I want to have, even in not full Dockerized projects, because local instances of SQL Server (and any database system) become problematic when time goes and different versions collides. I hope this end working well, it can be a great simplification for my development environment, and anyone else.

Really, if the Docker problems I having didn't exist, the benefits are clear to low the entry barrier for any contributor. I'm not being against Docker, I'm trying to get it working, fighting problems that arise; I just didn't find all solutions yet. There are some doubts I still have based on lack of experience, like how the debugger works on that environment, but that's an advanced thing to try later. I know that, at the initial proposal for Docker, I was a bit reluctant about the need of it since I had already experience with a complex cross platform Nodejs project, so for me was like no need for that encapsulation to make it works. Anyway, I understand the benefits and I open to it (specially regarding SQL set-up, it's a really big benefit on that specific case), that's why I keep spending time trying to fix the problems and I'm commenting about them, not to blame it, but to have all the details that can lead to solutions and some doc about troubleshooting for anyone else.

**Comment by IagoSRL** [comment]
A positive update on my findings here: as commented at the meeting, out of the time of this project and for my own learning, I set-up a blank project on Nodejs with Typeorm and a MySQL database, trying and understading Typeorm set-up, entities/models, javascript vs typescript APIs, query builder and migrations; specially migrations, after some issue I got it working.

OK, now the important. I just tried there a simple connection from Typeorm to my local SQLServer instance pointing to the Loconomics database, hardcoding the connection URL at code, and works. I should have tried this "minimal set-up" before for easier understanding of the problem, since I tried hardcoding the URL at the back-end config file too to guess if was a escaping problem without luck (as commented previously). Then I changed the setting to come from an environment variable, and it fails. This reinforce my opinion with "escaping" issues. Just noting that there is no modules messing with the setting at all, just a variable at the Bash shell and getting it from `process.env.SQLURL` at code, so is not a project's thing.

As first conclusion, Typeorm / Tedious are able to work with this 'named instance' set-up. Problem comes from how passing the environment variable to Nodejs at Windows.

I changed the shell: I work usually with Bash for Windows, where I got the error, but trying using CMD (the old Windows command line), the variable is read correctly and connects. Should have tried this before.
After, I tried to run the back-end from CMD and it still fails. Closer, but there is something more, I will keep researching.
